### DISCUSSION TITLE
Editor feature - multiple rotation/translation planes

### DISCUSSION THREAD

**Discussion by Ivoz** [ppy/osu#13021] [Ideas]
TL;DR: A mapper customizable "rotational plane" along which Ctrl+{J, H, >, <} work

Hi, not sure if someone wants to put work into backporting this for current osu!; but lazer would be nice :3

Currently Ctrl+H, Ctrl+J, Ctrl+>, Ctrl+< are foundational workhorse functions of many many mappers. Used to create symmetries across the X and Y axes, you can observe their effects cropping up amongst the vast majority of maps. They allow a lot of quick variation and organisation in a short time.

However, this also "creates" an somewhat artificial limitation on the creative patterns of maps: far more rarely do mappers put in the work to create mapping symmetries when using these shortcuts, that then *aren't* aligned to the traditional X/Y plane. Yes, it is quite possible: simply create your pattern, and then rotate the group of objects.

However because that is a two step process that must be reversed by opposite steps, making small changes to it while achieving the original symmetry(s) becomes substantially more tedious. So a lot of the time you do not see that time put in, and only traditional horizontal/vertical symmetries are observed the vast majority of the time.

I propose an editor option which would allow rotating the translational plane upon which the above actions operate. A bit of linear algebra matrices I believe makes the actual math involved not very complex.

When the plane is set to 0 degrees, obviously the editor would work just as currently. However, (as one option of UX to implement) the mapper could add other numbers of degrees to set the translational plane to. They could then toggle/select between their customised planes (15, 30, 45, 20 degrees might be common ones; often mappers will use these rotations of single or groups of objects anyway; or maybe they would like to use 13.66667 just to be different) and, for example, if a 15 degree plane was currently selected then when pressing Ctrl+H would effectively flip+rotate CCW 15 degrees the object over the center. You could have these degree options on a drop down, with editable entries, and cycle-able by shortcut(s) perhaps.

Heck, just writing this I can imagine having another option that would allow the rotational/translational "center" to be customised to an object's (or arbitrary) center instead of the normal map center (and then also, reset). (Again, another optimisation of process over designing a pattern in the map center and then having to move it to its final position, which also makes it more tedious to edit later).

I believe this sort of option(s), by allowing many more translational & rotational results to be easily and quickly achieved while mapping (with functional shortcuts available), would lead to relatively quick, effective and appreciable expansion in the practical design methods and patterns of osu maps.

If I have not explained myself sufficiently and you would like to clarify anything don't hesitate to ask for such.

**Comment by pzrc** [comment]
ngl with how the editor is planned to work with rotations in the future I just find this to be a really unnecessary feature. diagonal symmetry exists in modern mapping and the process for doing it isn't inefficient as it is currently, so adding optional planes for reflection is just... bloat imo

**Comment by camellirite** [comment]
where are the plans for the editor? I've never seen them but it's not hard to do these things in the current editor

**Comment by Ivoz** [comment]
It is indeed not hard to _create_ a pattern with any rotation or translation, but it can be comparatively incredibly tedious to then **edit** a part of that pattern while preserving its previous properties you had when you made it.

That's the whole point of the proposal - make editing such patterns simple after you have adjusted the editor plane to what they were using. And creating them should be slightly faster as well, if for instance you were using an altered angle plane of symmetry for a lot of patterns or altered center.

**Comment by pzrc** [comment]
Fair. If the plane that is created persists, I'm all for it, considering the fact that you currently need to redo the pattern anyway if you mess up, which is really tedious 😊
